python/ray/autoscaler/aws/config.py

```diff
@@ -359,10 +378,27 @@ def _get_key(key_name, config):
 
 def _client(name, config):
     boto_config = Config(retries={"max_attempts": BOTO_MAX_RETRIES})
-    return boto3.client(name, config["provider"]["region"], config=boto_config)
+    try:
+        aws_credentials = config["provider"]["extra_config"]["aws_credentials"]
+        return boto3.client(
+            name,
+            config["provider"]["region"],
+            config=boto_config,
+            **aws_credentials)
+    except KeyError:
+        return boto3.client(
+            name, config["provider"]["region"], config=boto_config)
```

Comment:
We already assume that everything within the `provider` keyspace is custom to AWS (see the Azure, GCP `provider` fields), so the `extra_config` field is redundant.

Consider:

```python
    aws_credentials = config["provider"].get("aws_credentials", {})
    return boto3.client(
        name,
        config["provider"]["region"],
        config=boto_config,
        **aws_credentials)

Reply:
You are correct that everything within the `provider` keyspace is custom to AWS. 

Essentially, I want to pass in some configuration that is not specified in `ray-schema.json` (i.e. `aws_credentials`, `security_group` and etc), and my understanding is that we only allow properties that are listed in `ray-schema.json`  to be present in the config object. In other words, `config["provider"]["aws_credentials"]` is not allowed with the current schema.  I am using `extra-config` to pass them in because it's described as "provider-specific config" (see [here](https://github.com/ray-project/ray/blob/37e4d29f8715330eb772f05513c696fd852ab565/python/ray/autoscaler/ray-schema.json#L131)) , which I thought was appropriate for the use case.

Reply:
As for how to get the value from the config object, I debated between what I currently have and what you purposed. If am correct about using `extra_config`, using the approach you proposed we would get some like 

`config["provider"].get("extra_config", {}).get("security_group", {}).get("IpPermissions", {})` for getting `IpPermissions`.

I personally think the KeyError way is easier to understand in this case, but if you feel strongly about avoiding it I'm happy to change that.

Reply:
Ah - let's change that. The schema was added last week, and for provider configurations, there's no particular reason why it needs to be stiff.

```diff
diff --git a/python/ray/autoscaler/ray-schema.json b/python/ray/autoscaler/ray-schema.json
index 81bd1663a..c0b64a955 100644
--- a/python/ray/autoscaler/ray-schema.json
+++ b/python/ray/autoscaler/ray-schema.json
@@ -58,7 +58,7 @@
             "type": "object",
             "description": "Cloud-provider specific configuration.",
             "required": [ "type" ],
-            "additionalProperties": false,
+            "additionalProperties": true,
             "properties": {
                 "type": {
                     "type": "string",
Stage this hunk [y,n,q,a,d,j,J,g,/,e,?]? y
@@ -128,10 +128,6 @@
                     "type": "object",
                     "description": "k8s autoscaler permissions, if using k8s"
                 },
-                "extra_config": {
-                    "type": "object",
-                    "description": "provider-specific config"
-                },
                 "cache_stopped_nodes": {
                     "type": "boolean",
                     "description": " Whether to try to reuse previously stopped nodes instead of launching nodes. This will also cause the autoscaler to stop nodes instead of terminating them. Only implemented for AWS."
```

Reply:
I agree that chaining 3 gets is not practical. If we can remove the `extra_config` layer, that should make your life easier.

Reply:
^ feel free to make this above change.

Reply:
Made the changes as you suggested.